File: milestone 2/series4/sde_convstrongmil.m
function sde_convstrongmil
% SDE X_t = f(X_t,t)dt + g(X_t,t)dB_t
%     X_0 = x0; t in [0,T] is solved using the EM method
%
% all paths are computed at the same time
N0=10;                            % number of steps on coarsest level
L=6;                              % number of refinement steps
M=10^4;                           % number of samples 

T = 1;                            % final time
x0 = 1;                           % initial condition
f=@(x,t)(-sin(x).*(cos(x)).^3);   % coeffients of the SDE
g=@(x,t)cos(x).^2;
hfu=@(x,t)2*cos(x).^3.*(-sin(x));

NL = N0*2^L;
B = brownp(T,NL,M);               % paths of Brownian motion on finest level
YT = zeros(L+1,M);                % values of X_T for h=T/(N*2^l), l=0,...,L, all m
tic
for l=0:L                         % for N = N0, N0*2, ... , N0*2^L
  N = N0*2^l;
  p = 2^(L-l);                    % p = NL/N
  h = T/N; 
  Y = zeros(N+1,M);
  Y(1,:) = x0;
  for j=1:N                       % perform N steps of E-M method:
    dB = B(1+j*p,:) - B(1+(j-1)*p,:);
    t = j*h;
    x=Y(j,:);
Y(j+1,:) = x + f(x,t)*h + g(x,t).*dB;
Y(j+1,:) = x + f(x,t)*h + g(x,t).*dB+0.5*hfu(x,t).*(dB.^2-h);
  end
  YT(l+1,:) = Y(N+1,:);           % values of X_T
end

%strong convergence :              
exact=atan(B(end,:)+tan(x0));
YTe = abs(YT-repmat(exact,L+1,1));   %    errors for XT
YTem = mean(YTe,2);             %    mean errors for strong convergence
YTem2 = sum(YTe.^2,2).^0.5;         %    mean square errors for strong convergence

Lp = L;                       % plot results for l=0,...,Lp 
                          
hv = T./(N*2.^(0:Lp)');
hL = hv(end); p = hv(1)/hL;

%figure(1); loglog(hv,YTem2,'-gx',hv,YTem,'-ro',hL*[1 p],YTem(end)*[1 p^.5],hL*[1 p],YTem2(end)*[1 p^.5]);
figure(1); loglog(hv,YTem2,'-gx',hv,YTem,'-ro',hL*[1 p],YTem(end)*[1 p^1],hL*[1 p],YTem2(end)*[1 p^1]); 
legend('strong error in L^1', 'strong error in L^2', ' h^{1}',' h^{1}'); grid on
xlabel('step size $h$')
ylabel('error')
% compute convergence rate
p = polyfit(log(hv),log(YTem),1);
disp('Strong rate of convergence')
disp(p(1));

toc

File: milestone 2/series4/sde_convstrongmil.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import linregress
from time import time
from brownp import brownp

def sde_convstrongmil():
    N0 = 10
    L = 6
    M = 10**4

    T = 1
    x0 = 1
    f = lambda x, t: -np.sin(x) * (np.cos(x))**3
    g = lambda x, t: np.cos(x)**2
    hfu = lambda x, t: 2 * np.cos(x)**3 * (-np.sin(x))

    NL = N0 * 2**L
    B = brownp(T, NL, M)
    YT = np.zeros((L+1, M))

    start_time = time()

    for l in range(L+1):
        N = N0 * 2**l
        p = 2**(L-l)
        h = T / N
        Y = np.zeros((N+1, M))
        Y[0, :] = x0

        for j in range(1, N+1):
            dB = B[j*p, :] - B[(j-1)*p, :]
            t = j * h
            x = Y[j-1, :]
            Y[j, :] = x + f(x, t) * h + g(x, t) * dB
            Y[j, :] = x + f(x, t) * h + g(x, t) * dB + 0.5 * hfu(x, t) * (dB**2 - h)

        YT[l, :] = Y[N, :]

    exact = np.arctan(B[-1, :] + np.tan(x0))
    YTe = np.abs(YT - np.tile(exact, (L+1, 1)))
    YTem = np.mean(YTe, axis=1)
    YTem2 = np.sqrt(np.sum(YTe**2, axis=1))

    Lp = L
    hv = T / (N * 2 ** np.arange(0, Lp+1))
    hL = hv[-1]
    p = hv[0] / hL

    plt.loglog(hv, YTem2, '-gx', hv, YTem, '-ro', hL * np.array([1, p]), YTem[-1] * np.array([1, p**1]), hL * np.array([1, p]), YTem2[-1] * np.array([1, p**1]))
    plt.legend(['strong error in L^1', 'strong error in L^2', 'h^{1}', 'h^{1}'])
    plt.grid(True)
    plt.xlabel('step size $h$')
    plt.ylabel('error')
    plt.show()

    slope, _, _, _, _ = linregress(np.log(hv), np.log(YTem))
    print('Strong rate of convergence:', slope)

    elapsed_time = time() - start_time
    print('Elapsed time:', elapsed_time)
sde_convstrongmil()